docs/assigning-global-variables.md

```diff
@@ -0,0 +1,199 @@
+  globalThis[globalName] = contextless;
+}
+```
+
+You can then write above example as:
+
+```typescript
+class EventHandlers {
+  @registerEventHandler
+  public OnStart() {
+    // start event handling code
+  }
+  @registerEventHandler
+  public OnStateChanged(newState: State) {
+    // state change event handler code
+  }
+}
+```
+
+:::note
+In the above example, `this` will be `nil` in the methods, do not try to use other members in the EventHandlers class!
+:::
+
+## Registering classes with a decorator
+
+Sometimes you want to register classes as globals, you can also do that with a decorator:
+
+```typescript
+function registerClass<TClass, TArgs extends unknown[]>(
+  c: new (...args: TArgs) => TClass,
+  context: ClassDecoratorContext,
+) {
+  if (context.name) {
+    // @ts-ignore tell TS to ignore us 'illegally' writing to global scope
+    globalThis[context.name] = c;
+  }
+}
+```
+
+You can now register any class by simply adding the decorator:
+
+```typescript
+@registerClass
+class EventHandlers {
+  public OnStart() {
+    // start event handling code
+  }
+  public OnStateChanged(newState: State) {
+    // state change event handler code
+  }
+}
+```
+
+### Custom global name decorator
+
+In the examples above, the decorators directly used the name of the decorated class or method, but with decorator parameters you can also specify custom override names:
+
+```typescript
+const registerClass =
+  (globalName: string) =>
+  <TClass, TArgs extends unknown[]>(c: new (...args: TArgs) => TClass, context: ClassDecoratorContext) => {
+    if (context.name) {
+      // @ts-ignore tell TS to ignore us 'illegally' writing to global scope
+      globalThis[globalName] = c;
+    }
+  };
+```
+
+Now instead of taking the global name from the class, you can specify a custom name yourself:
+
+```typescript
+@registerClass("CustomGlobalName")
+class EventHandlers {
+  public OnStart() {
+    // start event handling code
+  }
+  public OnStateChanged(newState: State) {
+    // state change event handler code
+  }
+}
+```
+
+## Assigning to globals with declarations
+
+The main weakness of the above methods is that you can declare any string, not protecting you from typos, and not giving any kind of editor support. This is fine if there are only a few such registrations that need to be done, but is somewhat error prone.
+
+An alternative method would be to explicitly declare the global variables in a declarations file:
+
+```ts
+declare var OnStart: (this: void) => void;
+declare var OnStateChanged: (this: void, newState: State) => void;
+```
+
+You can then assign to these functions as if they were global variables:
+
+```typescript
+OnStart = () => {
+  // start event handling code
+};
+OnStateChanged = (newState: State) => {
+  // state change event handler code
+};
+```
+
+This of course only works if you know the names of the global variables beforehand, if these names are dynamic, consider using one of the other methods instead.
```

Comment:
I think this should be the first option.

As this is the "most correct/safe" option. The other ones maybe more convenient, but also slightly more hacky.